src/herder/Upgrades.h

```diff
@@ -20,34 +20,66 @@ class Database;
 struct LedgerHeader;
 struct LedgerUpgrade;
 
-class Upgrades
+class UpgradeParameters
 {
+    VirtualClock::time_point mUpgradeTime;
```

Comment:
The only thing I'm a little nervous about in this PR is that you're relying on the users of the UpgradeParameters class to check cases in a specific order: first call `if(upgradeAtLedger())` and only proceed to use `getUpgradeAtTime()` if the former call returns false. This is a bit accident-prone: someone could fail to check in this order and wind up accessing a (meaningless but initialized and valid) "upgrade time" in an instance that has only an upgrade ledger.

In general I'd prefer this type be made something that is more defensive about being in exactly one of two states -- time or ledger -- and prohibiting mixtures of both, or neither.